pages/blog/json-schema-beyond-validation.md

```diff
@@ -0,0 +1,152 @@
+---
+title: "Using JSON Schema beyond validation"
+date: 2021-07-27T12:00:00+01:00
+type: Engineering
+tags:
+  - tools
+  - code generation
+  - JSON Schema
+cover: /img/posts/jonaslagoni/json-schema-beyond-validation/halgatewood-com-QM9yzAoX-GQ-unsplash.webp
+authors:
+  - name: Jonas Lagoni
+    photo: /img/avatars/jonaslagoni.webp
+    link: https://github.com/jonaslagoni
+    byline: AsyncAPI Maintainer
+featured: true
+---
+
+What is JSON Schema, why is it important and why is it so hard to use beyond validation? 
+
+For those unfamiliar with Asyncapi [we use a superset of JSON Schema](https://www.asyncapi.com/docs/specifications/2.0.0#schemaObject) as the default format for defining operation payloads, headers, channel parameter schemas, etc.
+
+Even though we allow other formats such as Avro, OpenAPI 3.x and Swagger 2.x, RAML schemas in its place, as soon as it hits the parser (which most tooling utilizes), said formats are converted to [JSON Schema draft 7](https://json-schema.org/specification-links.html#draft-7) to ensure a [common structure for tooling](https://github.com/asyncapi/parser-js/blob/826b36922260254ba23d162cda309fc72f552c49/lib/models/message.js#L20). 
+
+However, in tooling, many times you do not want to validate data, but to represent the data in a structured manner so it is easier to interact with, such as classes that represent a message payload.
+
+## Quick intro to JSON Schema
+
+Let's try and take a look at an example. Given the following, we have a schema representing the validation rules that our data should comply with. 
+
+<figure>
+  <img src="../img/posts/jonaslagoni/json-schema-beyond-validation/json-schema-process.webp" title="JSON Schema validation process" alt="Shows the overall JSON Schema validation process of how a JSON Schema and some data is validated against each other."/>
+  <figcaption className="text-center text-gray-400 text-sm">Displays the overall process of validating data using JSON Schema.</figcaption>
+</figure>
+
+The JSON Schema defines that the JSON data should be an object, which requires a property called `someRequiredProperty` to always be present and an optional property called `someOptionalProperty`. `someRequiredProperty` should validate against an integer and `someOptionalProperty` against an arbitrary string. The schema also dictates that no additional properties (`"additionalProperties": false`) may be allowed. There is also some metadata defined, called `$id` and `$schema`, but they are not important for this example.
+
+If we then take a look at the example [data instances](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-4.2), the first one contains the required property, and the second one both the required and the optional property.
```

Comment:
are you referring to some specific example here, I clicked the link but I'm lost

Reply:
I see 😅 Reformulated the sentence to 
```
If we then take a look at the example data instances (below the JSON Schema in the figure), the first one contains the required property, and the second one has both the required and the optional property.

```

Reply:
oh, well done, I just misunderstood you link to an example and not that you want to provide a reference for people to understand what data instances is 😅